### PR TITLE
 uploading files #4314

### PR DESCRIPTION
**Pull request by naazy**
https://github.com/OxfordAbstracts/application/issues/4314 uploading files

### CODE REVIEW

File: src/ManageInformationView.elm [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                                    , type_ "text"
+                                    , placeholder "File title"
+                                    , id ("file-title-" ++ (toString f.id))
+                                    , onInput (ChangeFileToSaveTitle f.id)
+                                    ]
+                                    []
+                                ]
+                            , div [ class "form__question-sub-section form__question-sub-section--table" ]
+                                [ input
+                                    [ id (toString f.id)
```
Comment (roryc89) [inline src/ManageInformationView.elm]: Having and the DOM id as a common integer like "1" is likely to introduce sneaky bugs as another element could easily have the same id

File: src/ManageInformationView.elm [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                                    , class "button button--secondary icon icon--bin"
+                                    ]
+                                    []
+                                ]
+                            ]
+                    )
+
+        displayWarning =
+            if not (String.isEmpty (manageInformationWarning model)) then
+                True
```
Comment (roryc89) [inline src/ManageInformationView.elm]: can simply be `not (String.isEmpty (manageInformationWarning model))`

File: main.js [hunk 3]
```diff
@@ -62,5 +62,37 @@ function init(options, dev) {
       app.ports['deleteSession'].send(sessionId);
     }
   });
+// thanks to https://github.com/phylor/elm-image-upload/blob/master/index.html
```
Comment (roryc89) [inline main.js]: do we need more ports to do this? There is this [library](https://github.com/simonh1000/file-reader) that handles file upload without ports. Example here http://simonh1000.github.io/2016/12/elm-s3-uploads/

File: tests/dummy/DummyTypes.elm [hunk 4]
```diff
@@ -61,6 +61,11 @@ makeDummyModel sessions =
     , scheduleSubmissionsIndividually = False
     , showMobileView = False
     , displayedColumn = Nothing
+    , filesToSave = [ (FileToSave 0 "" "" "") ]
```
Comment (roryc89) [inline tests/dummy/DummyTypes.elm]: [extensible records](https://medium.com/@ckoster22/advanced-types-in-elm-extensible-records-67e9d804030d) can prevent this sort of upkeep.

File: src/ManageInformationView.elm [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                                    [ onClick (DeleteFileToSave f.id)
+                                    , class "button button--secondary icon icon--bin"
+                                    ]
+                                    []
+                                ]
+                            ]
+                    )
+
+        displayWarning =
+            if not (String.isEmpty (manageInformationWarning model)) then
```
Comment (roryc89) [inline src/ManageInformationView.elm]: can simply be `not (String.isEmpty (manageInformationWarning model))`

File: src/MainUpdate.elm [hunk 6]
```diff
@@ -1053,6 +1102,95 @@ update msg model =
...
+                    newFilesToSave =
+                        List.filter (\f -> f.id /= newFileId) model.filesToSave
+                in
+                    ( { model | filesToSave = List.append newFilesToSave [ newFile ], showSavingFilesSpinner = False }
+                    , Cmd.none
+                    )
+
+            FileSelected id ->
+                ( { model | showSavingFilesSpinner = True }
+                , fileSelected (toString id)
```
Comment (roryc89) [inline src/MainUpdate.elm]: Having small integers like "1" as DOM ids is dangerous as they can easily be repeated elsewhere in the page and introduce error.

File: src/ManageInformationView.elm (hunk 1 above)
Comment (naazy) [reply src/ManageInformationView.elm]: ah yes good point i was meant to change this :smile:

File: src/ManageInformationView.elm (hunk 5 above)
Comment (naazy) [reply src/ManageInformationView.elm]: 👍

File: main.js (hunk 3 above)
Comment (roryc89) [reply main.js]: From what I've read, using libraries native code seems to have pitfalls as well so no need actually!

File: tests/dummy/DummyTypes.elm (hunk 4 above)
Comment (naazy) [reply tests/dummy/DummyTypes.elm]: awesome good to know for future

Comment (nelsonic) [reply tests/dummy/DummyTypes.elm]: Great read @roryc89 hope you don't mind me adding it to: https://github.com/dwyl/learn-elm/issues/102 😉

Comment (roryc89) [reply tests/dummy/DummyTypes.elm]: @nelsonic of course! Share away! :)

**Pull request merged**